### PR TITLE
dev/financial#180 Fix line item calculation regression on line items (incorrectly treating as exclusive)

### PR DESCRIPTION
**Pull request by eileenmcnaughton**
Overview
=======================
Fixes a 5.41 regression where total amount was assumed to be exclusive when generating line items and calculating tax when it  is only sometimes exclusive

Before 
======================
Total amount is assumed to be tax-exclusive in all scenarios and line items calculated on that assumption

After 
=======================
Total amount is assumed to be inclusive in the bao - the edge case where it is exclusive historically and tested to be so is moved to contribution.create api

Technical Details
=======================

A [recent fix](https://github.com/civicrm/civicrm-core/commit/e967ce8fe2b58b94e2163dde395542e55599da13#diff-a16d4d7449cf5f3a0616d1d282a32f27ab6d3f7d2726d076c02ad1d4d655af41R512) added tax_amount calculations to getLineItems and used that amount to calculate the tax amount. 

However, in accordance with `api_v3_TaxContributionPageTest` I reached the conclusion that the 'normal behaviour' was to [treat the passed in amount as inclusive](https://github.com/civicrm/civicrm-core/commit/e967ce8fe2b58b94e2163dde395542e55599da13#diff-4c9d0b1abe07057a4eea2b47bc627eecb95face8ed8d86c1c005312a52cca811L4026-L4035). In fact that is historically the case ONLY [IF tax_amount is not passed in](https://github.com/civicrm/civicrm-core/commit/e967ce8fe2b58b94e2163dde395542e55599da13#diff-4c9d0b1abe07057a4eea2b47bc627eecb95face8ed8d86c1c005312a52cca811R187) - and it seems that it most commonly IS passed in. 

This moves the weird handling to the v3 contribution api. The intention is that v4 contribution api and order api will always assume that it is inclusive - although we can discuss adding the flexibility to define exclusive instead from the order api.

Note that when line_item (e.g as core forms do) is passed the bug is not hit -  the focus really is on the contribution.create api when it is allowing the BAO to calculate the line items.

Comments
=======================

@KarinG @mattwire @monishdeb @jitendrapurohit @christianwach this is the fix for the original sin - I think with this merged it will fix Contribution.api to 'do the right thing' when tax_amount is passed in. The fix for the order api becomes pretty simple at that point https://github.com/civicrm/civicrm-core/pull/21230/commits/05d1bbaade730155b18c419be5a090c2b5d2f220

However, I'm still expecting some tests that were altered AFTER the bug was introduced to fail with this or the order fix - just because the tests are wrong I think.

### CODE REVIEW

**Comment by monishdeb** [comment]
@eileenmcnaughton I can confirm that the patch correctly sets the line_item.line_total to be exclusive of tax_amount and is stored separately in line_item.tax_amount. Unlike for contribution.total_amount which is inclusive of tax_amount and there is special handling at Contribution.create api3. But there is a test failure https://test.civicrm.org/job/CiviCRM-Core-PR/43292/testReport/junit/(root)/api_v3_TaxContributionPageTest/testCreateUpdateContributionChangeTotal/ which is related and needs similar fix. 

As this patch is touching the essential part of financial code, I need anyone else to confirm my assessment.

File: tests/phpunit/api/v3/TaxContributionPageTest.php [hunk 1]
```diff
@@ -392,8 +392,8 @@ public function testCreateUpdateContributionChangeTotal(): void {
     ]);
 
     $this->assertEquals('300.00', $lineItems);
-    $this->assertEquals('320.00', $this->_getFinancialTrxnAmount($contribution['id']));
-    $this->assertEquals('320.00', $this->_getFinancialItemAmount($contribution['id']));
+    $this->assertEquals('300.00', $this->_getFinancialTrxnAmount($contribution['id']));
```
Comment (eileenmcnaughton) [inline tests/phpunit/api/v3/TaxContributionPageTest.php:395]: This reverts the changes made in the same commit as the original sin

**Comment by jitendrapurohit** [comment]
Tested the changes on the following scenarios -

**Frontend** - 

Online Contribution Page.
Online Membership Signup.
Online Membership Renewal.
Repeat the above test with priceset.

**Backend** -

New membership (w/o priceset)
Renew Membership 
New Membership with live mode payment (w/o priceset).
Renew Membership with live mode payment.

**Webform** -

Webform purchase of New Membership.
Webform donation with taxable FT.
If there are different financial type (both taxable) line items on a single webform, the tax amount is not saved correctly, but it is also an issue before this PR so unrelated.

It passes on all the above steps - expected values are present in line_item table and the amounts looks fine on view contribution and view membership pages.

I think we're good to merge this 👍 .

**Comment by KarinG** [comment]
@jitendrapurohit - re: webform and tax amount -> I think the better_line items branch addresses that (I use it in D7 and need to revamp it’s on our Google doc for D8/D9 port and inclusion).

**Comment by eileenmcnaughton** [comment]
@KarinG this fixes line_item.total_amount being incorrectly inclusive when calculated in the BAO - I'm not 100% sure if we need an order api fix on top of this

@jitendrapurohit I think you are saying
1) we should merge this
2) we need to do a further fix for the order api - I believe this is the extra fix - https://github.com/civicrm/civicrm-core/pull/21230 - note the failing tests there I believe have all been altered recently to use the order api rather than contribution api & added the incorrect assumption into them
3) we should back port this to 5.40 to ensure line_item.total_amount is not incorrectly inclusive

**Comment by eileenmcnaughton** [comment]
Also @jitendrapurohit I think this would mostly impact contribution.create api calls but not so much core forms - which generally pass in 'line_item' as a parameter to contribution.create

**Comment by eileenmcnaughton** [comment]
@monishdeb I think we should merge this based on @jitendrapurohit & then focus on #21230 which should address the originally identified bug

**Comment by monishdeb** [comment]
Agree. Merging now based on @jitendrapurohit @KarinG and my feedback.

**Comment by KarinG** [comment]
I'll add dev-master to our automated webform testing so that I can run this PR against all our tests.

**Comment by eileenmcnaughton** [comment]
Thanks @monishdeb @jitendrapurohit @KarinG 

I've rebased #21230 to just have the order api fix

I haven't updated the tests in it at this stage because I think it makes it possibly easier to check & agree the tests need updating

**Comment by KarinG** [comment]
I'm getting the same error using dev-master right now [when not passing in line_items] -> but no new errors:
image

But I can confirm -> https://github.com/civicrm/civicrm-core/pull/21189 dev/rc#14 handle api calls post schema change #21189 is fixed ✅

**Comment by eileenmcnaughton** [comment]
@KarinG  #21230 is the fix I think we should be merging to fix ^^

**Comment by KarinG** [comment]
Understood - just wanted to flag that no other issues surfaced on our end post merging this PR.

**Comment by eileenmcnaughton** [comment]
thanks @KarinG

**Pull request merged**